server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterService.java

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.health.HealthStatus;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * This service reports the health of master stability.
+ * If we have had a master within the last 30 seconds, and that master has not changed more than 3 times in the last 30 minutes, then
+ * this will report GREEN.
+ * If we have had a master within the last 30 seconds, but that master has changed more than 3 times in the last 30 minutes (and that is
+ * confirmed by checking with the last-known master), then this will report YELLOW.
+ * If we have not had a master within the last 30 seconds, then this will will report RED with one exception. That exception is when:
+ * (1) no node is elected master, (2) this node is not master eligible, (3) some node is master eligible, (4) we ask a master-eligible node
+ * to run this service, and (5) it comes back with a result that is not RED.
+ * Since this service needs to be able to run when there is no master at all, it does not depend on the dedicated health node (which
+ * requires the existence of a master).
+ */
+public class StableMasterService implements ClusterStateListener {
```

Comment:
As this would now be a standalone service, should we name this something a bit more descriptive (this service will look into discovery and cluster formation - and maybe expose some of those analyses via different public methods at some point, even though I do appreciate now it only has one public method)

Would `ClusterAnalysisService`/ `CoordinationAnalysisService` / `CoordinationDiagnosticsService` be a bit more descriptive? 
With a method `analyseMasterStability`/ `diagnoseMasterStability` instead of `calculate` ? 
In the same context, would `StableMasterDetails` be better named something along the lines of `CoordinationAnalysisDetails`?

These are all off-the-cuff suggestions (naming is, well, hard :) )

cc @DaveCTurner

Reply:
Sounds as good as anything. I'll go with that unless someone comes up with something better.